### ISSUE TITLE
Support disabling @Autowired on a per-@Bean basis [SPR-8396]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-framework#13043]
**[Stepan Koltsov](https://jira.spring.io/secure/ViewProfile.jspa?name=yozh)** opened **[SPR-8396](https://jira.spring.io/browse/SPR-8396?redirect=false)** and commented

```
public class AutowireOverrides {

    public static class TheBean {

        @Autowired
        @Value("from property")
        public void setBla(String value) {
            System.out.println("setBla: " + value);
        }
    }

    @Configuration
    public static class ContextConfiguration {

        @Bean
        public TheBean theBean() {
            TheBean r = new TheBean();
            r.setBla("from @Bean");
            return r;
        }

    }

    public static void main(String[] args) {
        new AnnotationConfigApplicationContext(ContextConfiguration.class);
    }

}
```

Obviously outputs

```
setBla: from @Bean
setBla: from property
```

and property specified in `@Bean` method has lower priority.

Requesting a `@Bean`(properties=...) attribute, so theBean() declaration could be written as:

```
@Bean(
  properties={
    @Property(name="bla", value="from @Bean")
  }
)
public TheBean theBean() {
    return new TheBean();
}
```

In this case Spring would call setBla only once with "from `@Bean`" parameter value.

Assume HttpServer library class with `@Autowired` threadPool property. I cannot declare two HttpServer instances that use different thread pools in the same application (without disabling autowire, switching to xmlconf etc).

---

**Affects:** 3.1 M1

**Attachments:**
- [mylyn-context.zip](https://jira.spring.io/secure/attachment/18343/mylyn-context.zip) (_10.84 kB_)

**Issue Links:**
- #20752 Support for autowire-candidate with `@Bean`

1 votes, 4 watchers

**Comment by jhoeller** [comment]
We have no intention to introduce a dedicated mechanism for such disabling.

Our recommendation is rather to structure application components accordingly between constructor arguments and bean properties: Constructor arguments are effectively replaced in the `@Bean` method body, potentially with a choice of several constructors, bypassing Spring's common constructor resolution (as with XML-defined beans) completely. Additional autowiring will still happen on the returned instance for bean properties and also for fields.

Alternatively, you may entirely go for constructor-based injection and declare the required autowired arguments as parameters on your `@Bean` method, passing them into the target constructor programmatically. This is arguably cleaner since it all happens within a single factory method then, composing the target state from factory method parameters and local state as needed.
